File: leetcode/easy/GuessNumberHigherOrLower.py
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if num is higher than the picked number
#          1 if num is lower than the picked number
#          otherwise return 0
# def guess(num: int) -> int:

class Solution:
    def guessNumber(self, n: int) -> int:
        # Initialize the search space for binary search.
        # The picked number is guaranteed to be between 1 and n, inclusive.
        low, high = 1, n

        # Loop until the search space is exhausted (low > high).
        while low <= high:
            # Calculate the middle of the current range.
            # This formula is used to prevent potential integer overflow in other languages,
            # though it's not strictly necessary in Python.
            middle = low + (high - low) // 2
            
            # Call the guess API to check if the middle number is the target.
            res = guess(middle)

            # If guess(middle) returns 0, we've found the number.
            if res == 0:
                return middle
            # If guess(middle) returns 1, our guess is too low,
            # so the target must be in the upper half of the range.
            elif res == 1:
                low = middle + 1
            # If guess(middle) returns -1, our guess is too high,
            # so the target must be in the lower half of the range.
            else:  # res == -1
                high = middle - 1
        

File: leetcode/easy/GuessNumberHigherOrLower.java
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * int guess(int num);
 */

/**
 * Solution class for the Guess Number Higher or Lower problem.
 * This class extends GuessGame which provides the guess(int num) API.
 */
public class Solution extends GuessGame {
    /**
     * Finds the secret number picked by the system using binary search.
     *
     * @param n The upper bound of the range in which the number is picked.
     * @return The secret number.
     */
    public int guessNumber(int n) {
        // Initialize the search space for binary search.
        // 'low' is the lower bound, and 'high' is the upper bound.
        int low = 1;
        int high = n;

        // Perform binary search until the search space is exhausted.
        while (low <= high) {
            // Calculate the middle of the current search range.
            // Using (low + (high - low) / 2) prevents potential integer overflow
            // compared to (low + high) / 2.
            int middle = low + (high - low) / 2;
            
            // Call the predefined guess API to get feedback on the middle number.
            int result = guess(middle);

            if (result == 0) {
                // The guess is correct. We found the number.
                return middle;
            } else if (result == 1) {
                // The guess is too low, so the secret number must be in the upper half.
                // Adjust the lower bound to search in the range [middle + 1, high].
                low = middle + 1;
            } else { // result == -1
                // The guess is too high, so the secret number must be in the lower half.
                // Adjust the upper bound to search in the range [low, middle - 1].
                high = middle - 1;
            }
        }

        // This line is unreachable given the problem constraints that the number
        // is always within the range [1, n]. It's included to satisfy the
        // method's signature which requires a return value.
        return -1;
    }
}